codegen/generate.go

```diff
@@ -1,26 +1,126 @@
 package codegen
 
 import (
+	"bufio"
+	"errors"
 	"fmt"
 	"os"
 	"path/filepath"
 	"strings"
+
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/vault/sdk/framework"
+)
+
+// generatedDirPerms uses 0775 because it is the same as for
+// the "vault" directory, which is at "drwxrwxr-x".
+const generatedDirPerms os.FileMode = 0775
+
+var (
+	errUnsupported = errors.New("code and doc generation for this item is unsupported")
+
+	// pathToHomeDir yields the path to the terraform-vault-provider
+	// home directory on the machine on which it's running.
+	// ex. /home/your-name/go/src/github.com/terraform-providers/terraform-provider-vault
+	pathToHomeDir = func() string {
+		repoName := "terraform-provider-vault"
+		wd, _ := os.Getwd()
+		pathParts := strings.Split(wd, repoName)
+		return pathParts[0] + repoName
+	}()
 )
 
-// pathToHomeDir yields the path to the terraform-vault-provider
-// home directory on the machine on which it's running.
-// ex. /home/your-name/go/src/github.com/terraform-providers/terraform-provider-vault
-var pathToHomeDir = func() string {
-	repoName := "terraform-provider-vault"
-	wd, _ := os.Getwd()
-	pathParts := strings.Split(wd, repoName)
-	return pathParts[0] + repoName
-}()
+func Run(logger hclog.Logger, paths map[string]*framework.OASPathItem) error {
+	// Read in the templates we'll be using.
+	h, err := newTemplateHandler(logger)
+	if err != nil {
+		return err
+	}
+	// Use a file creator so the logger can always be available without having
+	// to awkwardly pass it in everywhere.
+	fCreator := &fileCreator{
+		logger:          logger,
+		templateHandler: h,
+	}
+	createdCount := 0
+	for endpoint, templateType := range endpointRegistry {
+		logger.Info(fmt.Sprintf("generating %s for %s\n", templateType.String(), endpoint))
+		if err := fCreator.GenerateCode(endpoint, paths[endpoint], templateType); err != nil {
+			if err == errUnsupported {
+				logger.Warn(fmt.Sprintf("couldn't generate %s, continuing", endpoint))
+				continue
+			}
+			logger.Error(err.Error())
+			os.Exit(1)
+		}
+		// TODO in separate PR - add fCreator.GenerateDoc() method
+		createdCount++
+	}
+	logger.Info(fmt.Sprintf("generated %d files\n", createdCount))
+	return nil
+}
+
+type fileCreator struct {
+	logger          hclog.Logger
+	templateHandler *templateHandler
+}
+
+// GenerateCode is exported because it's the only non-internal method on the fileCreator.
```

Comment:
Oh, thanks for pointing that out! :-) I'll rephrase it to help it hopefully make sense.

Reply:
Rephrased it to:
```
// GenerateCode is exported because it's the only method intended to be used by
// other objects. Unexported methods may be available to other code in this package,
// but they're not intended to be used by anything but the fileCreator.
```